Use an init container to deliver the content

jhrozek opened the pull request:
Instead of baking the content into the scanner container, let's use a
separate init container with the content instead. This would allow us to
separate the scanner lifecycle from the content lifecycle and also allow
users to provide their own content.

The init container expects the content to be at the root of the
filesystem and all files matching \*.xml would be copied from the init
container's / to emptyDir volume mounted at /content by the scanner.

For development, the content image can be passed with an environment
variable:
    make run CONTENT_IMAGE=quay.io/jhrozek/ocp4-openscap-content:rhel-8

It might be useful to also allow specifying the content image in the
CRD, but let's do that in a subsequent patch.

jhrozek commented:
CC @JAORMX @mrogers950 

Here's one question I have: Should we already extend the CRD to include an override for the image and an override for the pull policy? Currently the pullPolicy is Always which I think is a reasonable default as the scans wouldn't be too frequent I guess, but maybe someone might be irritated about the pods pulling the image over and over again?

JAORMX commented:
I think leaving it as Always is good. And yes, lets update the CRD.

JAORMX commented on README.md:
```diff
@@ -105,6 +105,8 @@ Should you wish to override any of the two container images in the pod, you can
     * `OPENSCAP_IMAGE` for the scanner container
     * `LOG_COLLECTOR_IMAGE` for the log collecting container
+    * `CONTENT_IMAGE` for the initContainer that provides the content for the scanner. 
+      The image must contain the `cp` utility and the profiles must reside at `/`.
```
Lets make this part of the CRD instead.

jhrozek replied:
OK, thanks for the review, I'll do that.

jhrozek replied:
Done in the latest patches

jhrozek commented:
So the `operator-sdk generate k8s` command changed quite a few more things than I expected. First, a new CRD was generated (`deploy/crds/complianceoperator.compliance.openshift.io_compliancescans_crd.yaml`) so I removed the older one as well. The `zz_generated.openapi.go` file also changed some references from absolute (`github.com/jhrozek/compliance-operator`) to relative (`./pkg/apis/complianceoperator/v1alpha1.ComplianceScanStatus`). I attribute that to newer operator-sdk version. Let me do an in-cluster test as well, though..

jhrozek commented:
My in-cluster smoke test succeeded btw.

JAORMX commented on pkg/controller/compliancescan/compliancescan_controller.go:
```diff
@@ -401,3 +429,33 @@ func getOscapContainerEnv(scanSpec *complianceoperatorv1alpha1.ComplianceScanSpe
+		"always":       corev1.PullAlways,
+		"never":        corev1.PullNever,
+		"ifnotpresent": corev1.PullIfNotPresent,
+	}
```
Let's do this from the types.go side instead of the operator. You could probably pull in the pullpolicy from the corev1 API. Validation could be done in the CRD as well.

jhrozek commented:
On Wed, Nov 13, 2019 at 06:52:59AM -0800, Juan Osorio Robles wrote:
> JAORMX requested changes on this pull request.
> 
> 
> 
> > +	image := DefaultContentContainerImage
> +
> +	if scanSpec.ContentImage != "" {
> +		image = scanSpec.ContentImage
> +	}
> +
> +	logger.Info("Content image", "image", image)
> +	return image
> +}
> +
> +func getInitContainerPullPolicy(scanSpec *complianceoperatorv1alpha1.ComplianceScanSpec, logger logr.Logger) corev1.PullPolicy {
> +	var pullPolicyMapping = map[string]corev1.PullPolicy{
> +		"always":       corev1.PullAlways,
> +		"never":        corev1.PullNever,
> +		"ifnotpresent": corev1.PullIfNotPresent,
> +	}
> 
> Let's do this from the types.go side instead of the operator. You could probably pull in the pullpolicy from the corev1 API. 

Sure, just to be clar, the CRD would then directly contain corev1.PullPolicy?

> Validation could be done in the CRD as well.

Yeah, I saw that you can do stuff like:
```
// +kubebuilder:validation:Enum=foo;bar;baz
```
let me see about using the type from corev1 first..

jhrozek commented:
On Wed, Nov 13, 2019 at 07:01:05AM -0800, Jakub Hrozek wrote:
> On Wed, Nov 13, 2019 at 06:52:59AM -0800, Juan Osorio Robles wrote:
> > JAORMX requested changes on this pull request.
> > 
> > 
> > 
> > > +	image := DefaultContentContainerImage
> > +
> > +	if scanSpec.ContentImage != "" {
> > +		image = scanSpec.ContentImage
> > +	}
> > +
> > +	logger.Info("Content image", "image", image)
> > +	return image
> > +}
> > +
> > +func getInitContainerPullPolicy(scanSpec *complianceoperatorv1alpha1.ComplianceScanSpec, logger logr.Logger) corev1.PullPolicy {
> > +	var pullPolicyMapping = map[string]corev1.PullPolicy{
> > +		"always":       corev1.PullAlways,
> > +		"never":        corev1.PullNever,
> > +		"ifnotpresent": corev1.PullIfNotPresent,
> > +	}
> > 
> > Let's do this from the types.go side instead of the operator. You could probably pull in the pullpolicy from the corev1 API. 
> 
> Sure, just to be clar, the CRD would then directly contain corev1.PullPolicy?
> 
> > Validation could be done in the CRD as well.
> 
> Yeah, I saw that you can do stuff like:
> ```
> // +kubebuilder:validation:Enum=foo;bar;baz
> ```
> let me see about using the type from corev1 first..

After talking to @JAORMX on Slack, I've dropped the policy completely
and just use PullAlways. We can always add the policy, but let's not
complicate things unless we need to.

jhrozek commented:
Let's close this in favor of openshift/compliance-operator
